_posts/2020-07-23-demo-fraud-detection-3.md

```diff
@@ -0,0 +1,655 @@
+    that are not serializable and cannot be transferred from the
+    JobManager's JVM.
+
+Most importantly, `ProcessFunction` also has access to the fault-tolerant
+state, handled by Flink. This combination, together with Flink\'s
+message processing and delivery guarantees, makes it possible to build
+resilient event-driven applications with almost arbitrarily
+sophisticated business logic. This includes creation and processing of
+custom windows with state.
+
+### Implementation
+
+#### State and Clean-up
+
+In order to be able to process time windows, we need to keep track of
+data belonging to the window inside of our program. To ensure that this
+data is fault-tolerant and can survive failures in a distributed system,
+we should store it inside of Flink-managed state. As the time
+progresses, we do not need to keep all previous transactions. According
+to the sample rule, all events that are older than 24 hours become
+irrelevant. We are looking at a window of data that constantly moves and
+where stale transactions need to be constantly moved out of scope (in
+other words, cleaned up from state).
+
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/patterns-blog-3/window-clean-up.png" width="400px" alt="Figure 3: Window Clean-up"/>
+<br/>
+<i><small>Figure 3: Window Clean-up</small></i>
+</center>
+<br/>
+
+
+We will
+[use](https://ci.apache.org/projects/flink/flink-docs-stable/dev/stream/state/state.html#using-keyed-state)
+`MapState` to store the individual events of the window. In order to allow
+efficient clean-up of the out-of-scope events, we will utilize event
+timestamps as the `MapState` keys.
+
+In a general case, we have to take into account the fact that there
+might be different events with exactly the same timestamp, therefore
+instead of individual Transaction per key(timestamp) we will store sets.
+
+```java
+MapState<Long, Set<Transaction>> windowState;
+```
+
+<div class="alert alert-info" markdown="1">
+<span class="label label-info" style="display: inline-block"><span class="glyphicon glyphicon-info-sign" aria-hidden="true"></span> Side Note </span>
+when any Flink-managed state is used inside a
+`KeyedProcessFunction`, the data returned by the `state.value()` call is
+automatically scoped by the key of the *currently-processed event*
+- see Figure 4. If `MapState` is used, the same principle applies, with
+the difference that a `Map` is returned instead of `MyObject`. If you are
+compelled to do something like
+`mapState.value().get(inputEvent.getKey())`, you should probably be using
+`ValueState` instead of the `MapState`. As we want to store *multiple values
+per event key*, in our case, `MapState` is the right choice.
+</div>
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/patterns-blog-3/keyed-state-scoping.png" width="800px" alt="Figure 4: Keyed State Scoping"/>
+<br/>
+<i><small>Figure 4: Keyed State Scoping</small></i>
+</center>
+<br/>
```

Comment:
nice excursion!

Do you think, this figure should rather be inside the info box?